onnxruntime/core/providers/rocm/rocm_stream_handle.cc

```diff
@@ -0,0 +1,173 @@
+  }
+
+  hipEvent_t event_;
+};
+
+RocmStream::RocmStream(hipStream_t stream,
+                       const OrtDevice& device,
+                       AllocatorPtr cpu_allocator,
+                       bool release_cpu_buffer_on_rocm_stream,
+                       bool own_flag,
+                       miopenHandle_t external_miopen_handle,
+                       rocblas_handle external_rocblas_handle) : Stream(stream, device),
+                                                                own_stream_(own_flag),
+                                                                cpu_allocator_(cpu_allocator),
+                                                                release_cpu_buffer_on_rocm_stream_(release_cpu_buffer_on_rocm_stream) {
+  if (own_flag) {
+    ROCBLAS_CALL_THROW(rocblas_create_handle(&rocblas_handle_));
+    ROCBLAS_CALL_THROW(rocblas_set_stream(rocblas_handle_, stream));
+    MIOPEN_CALL_THROW(miopenCreate(&miopen_handle_));
+    MIOPEN_CALL_THROW(miopenSetStream(miopen_handle_, stream));
+  } else {
+    rocblas_handle_ = external_rocblas_handle;
+    ROCBLAS_CALL_THROW(rocblas_set_stream(rocblas_handle_, stream));
+    miopen_handle_ = external_miopen_handle;
+    MIOPEN_CALL_THROW(miopenSetStream(miopen_handle_, stream));
+  }
+}
+
+RocmStream::~RocmStream() {
+  ORT_IGNORE_RETURN_VALUE(CleanUpOnRunEnd());
+  if (own_stream_) {
+    rocblas_destroy_handle(rocblas_handle_);
+    miopenDestroy(miopen_handle_);
+    if (handle_)
+      HIP_CALL_THROW(hipStreamDestroy(static_cast<hipStream_t>(handle_)));
+  }
+}
+
+std::unique_ptr<synchronize::Notification> RocmStream::CreateNotification(size_t /*num_consumers*/) {
+  return std::make_unique<RocmNotification>(*this);
+}
+
+void RocmStream::Flush() {
+  if (own_stream_)
+    HIP_CALL_THROW(hipStreamSynchronize(static_cast<hipStream_t>(handle_)));
+}
+
+void RocmStream::EnqueDeferredCPUBuffer(void* cpu_buffer) {
+  // stream is per thread, so don't need lock
+  deferred_cpu_buffers_.push_back(cpu_buffer);
+}
+
+struct CpuBuffersInfo {   // TODO: should be moved to base class
+  AllocatorPtr allocator;
+  void** buffers;
+  // CPU buffer buffers[i].
+  // Number of buffer points in "buffers".
+  size_t n_buffers;
+};
+
+static void ReleaseCpuBufferCallback(hipStream_t /*stream*/, hipError_t /*status*/, void* raw_info) {  // TODO: should be moved to base class
+  auto info = reinterpret_cast<CpuBuffersInfo*>(raw_info);
+  for (size_t i = 0; i < info->n_buffers; ++i) {
+    info->allocator->Free(info->buffers[i]);
+  }
+  delete[] info->buffers;
+  delete info;
+}
+
+Status RocmStream::CleanUpOnRunEnd() {
+  if (deferred_cpu_buffers_.empty())
+    return Status::OK();
+  // Release the ownership of cpu_buffers_info so that the underlying
+  // object will keep alive until the end of ReleaseCpuBufferCallback.
+  if (release_cpu_buffer_on_rocm_stream_ && cpu_allocator_->Info().alloc_type == OrtArenaAllocator) {
+    auto cpu_buffers_info = new CpuBuffersInfo;
```

Comment:
>auto cpu_buffers_info = new CpuBuffersInfo; [](http://example.com/codeflow?start=4&length=43)

"new" here violate ort compliance, pls always be with smart pointers.

Reply:
Will do the change on cuda part first

Reply:
The code is moved from here in main branch:
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L381

The implementation is updated in main branch, we need to re-sync, in both cuda and AMD ep.